Router pim config

nabahr opened the pull request:
Added new 'router pim[6] [vrf NAME]' config node.

Moved all existing global/vrf PIM configs to the new subnode and the existing configuration was set to be hidden and deprecated.
Both versions of the configuration still work together for the time being until they can be safely removed.

PIM now always writes out configuration in the "router pim" syntax so ,in order to support frr-reload, the JSON based configuration in the PIM topotests also needs to generate configuration in the same syntax.

Config file style tests are left in the global pim configuration syntax for now as a test of backwards compatibility.

Fixes #16196

eqvinox commented:
oh god that diff size T_T …

eqvinox commented on pimd/pim_cmd_common.h:
```diff
@@ -190,4 +192,47 @@ int pim_show_bsr_helper(const char *vrf, struct vty *vty, bool uj);
+ * that VTY_CURR_XPATH is correct and/or uses relative paths, and at
+ * the end vty is moved back to the original node and xpath popped
+ */
+#define START_PIM_DEPRECATED(clinode)                           \
```
When and how are we going to get rid of this?

nabahr replied:
Well, it could go away now, but I'd say once it's been long enough that no body remembers the previous PIM syntax then this stuff can be removed.

eqvinox commented on pimd/pim_vty.c:
```diff
@@ -172,89 +172,66 @@ int pim_global_config_write_worker(struct pim_instance *pim, struct vty *vty)
 	struct pim_ssm *ssm = pim->ssm_info;
-	char spaces[10];
 
-	if (pim->vrf->vrf_id == VRF_DEFAULT)
```
is the indentation off now, for router pim blocks inside VRFs?

nabahr replied:
VRF-aware PIM configuration is also at the top level.
```
router pim
  ...
router pim vrf blue
  ...
```

nabahr commented:
> oh god that diff size T_T …

Yeah. I tried different ways to ALIAS the commands but couldn't get it to work so I essentially had to duplicate all of the global PIM commands. Also all of the pim global xpath's have been touched, since now the current XPath is set and commands use relative paths.
I am happy to hear any other ideas.

Jafaral commented:
Documentation needs to be updated to reflect the new syntax.

nabahr commented:
Verify source still fails for these 2 issues that won't be fixed

> < WARNING: Comparisons should place the constant on the right side of the test
< #5618: FILE: /tmp/f1-589488/pim_cmd_common.c:5618:

This is a simple comparison of two defines, either order produces this warning.

> < WARNING: Macros with flow control statements should be avoided
< #207: FILE: /tmp/f1-589488/pim_cmd_common.h:207:

This macro is meant to set up a few function variables and set the correct XPath for the deprecated CLI handlers. I can remove the macro if required but that'll require copying this duplicate code to all of the deprecated CLI handlers. As it is now I can't fix this warning.

Jafaral commented on pimd/pim_cmd_common.c:
```diff
@@ -5705,3 +5600,34 @@ int pim_show_bsm_db_helper(const char *vrf, struct vty *vty, bool uj)
+			continue;
+
+		snprintfrr(framestr, sizeof(framestr), "router pim%s",
+			   (PIM_AF == AF_INET6 ? "6" : ""));
```
Common code between pim and pim6. `PIM_AF == AF_INET6 ` triggers 
```
WARNING: Comparisons should place the constant on the right side of the test
```
The warning is trigged regardless  of the order, a rare use case.

nabahr commented:
Added a fix to frr-reload to detect and convert legacy pim configuration so that the tool works as expected. I reverted the changes to topotests so they are all back to using the legacy pim configuration.

riw777 reviewed (approved):
looks good
